File: bench/go/matmul.go
// bench/go/matmul.go — Go 版矩阵乘法（分块，与 bench/myp/matmul.myp 同算法）
// verify <迹>（浮点，容差对比）；与 C++/MYP 同算法。
package main

import (
	"fmt"
	"time"
)

func matmulTrace(n int) float64 {
	A := make([]float64, n*n)
	B := make([]float64, n*n)
	C := make([]float64, n*n)
	for i := 0; i < n*n; i++ {
		A[i] = float64(i % 1000)
		B[i] = float64(i % 7)
	}
	const BS = 64
	for i0 := 0; i0 < n; i0 += BS {
		for j0 := 0; j0 < n; j0 += BS {
			for k0 := 0; k0 < n; k0 += BS {
				for i1 := i0; i1 < i0+BS && i1 < n; i1++ {
					for k1 := k0; k1 < k0+BS && k1 < n; k1++ {
						av := A[i1*n+k1]
						for j1 := j0; j1 < j0+BS && j1 < n; j1++ {
							C[i1*n+j1] += av * B[k1*n+j1]
						}
					}
				}
			}
		}
	}
	trace := 0.0
	for i := 0; i < n; i++ {
		trace += C[i*n+i]
	}
	return trace
}

func main() {
	const n = 512
	t0 := time.Now()
	tr := matmulTrace(n)
	ms := time.Since(t0).Milliseconds()
	fmt.Printf("verify %.10g\nms %d\n", tr, ms)
}

File: bench/cpp/matmul.cpp
// bench/cpp/matmul.cpp — C++ 版矩阵乘法（与 bench/myp/matmul.myp 同算法）
// 输出：verify <迹> / ms <毫秒>
#include <chrono>
#include <cstdio>
#include <vector>

static double matmulTrace(int n) {
    std::vector<double> A(n * n), B(n * n), C(n * n, 0.0);
    for (int i = 0; i < n * n; ++i) {
        A[i] = (i % 1000);
        B[i] = (i % 7);
    }
    for (int r = 0; r < n; ++r)
        for (int c = 0; c < n; ++c) {
            double sum = 0.0;
            for (int k = 0; k < n; ++k) sum += A[r * n + k] * B[k * n + c];
            C[r * n + c] = sum;
        }
    double trace = 0.0;
    for (int i = 0; i < n; ++i) trace += C[i * n + i];
    return trace;
}

int main() {
    const int n = 512;
    auto t0 = std::chrono::steady_clock::now();
    double tr = matmulTrace(n);
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("verify %.10g\nms %.0f\n", tr, ms);
    return 0;
}
